File: 力扣/左/并查集/岛屿数量.cpp
#include<bits/stdc++.h>
using namespace std;

class UnionFind
{
public:
    UnionFind(int n)
    {
        p.resize(n);
        size.resize(n, 1);
        for (int i = 0; i < n; ++i)
            p[i] = i;
    }

    int find(int x)
    {
        if (p[x] != x)
            p[x] = find(p[x]);
        return p[x];
    }

    void unite(int x, int y)
    {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX != rootY)
        {
            if (size[rootX] < size[rootY])
                swap(rootX, rootY);
            p[rootY] = rootX;
            size[rootX] += size[rootY];
        }
    }
private:
    vector<int> p;
    vector<int> size;
};

int dx[4] = {0, 1, 0, -1};
int dy[4] = {1, 0, -1, 0};



class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        UnionFind unite(m * n + 1);
        

        for(int i = 0;i < m;i++)
        {
            for(int j = 0; j < n;j++)
            {
                if(grid[i][j] == '0')
                    continue;
                int x = i*n + j;

                for(int k = 0; k < 4;k++)
                {
                    int nx = i + dx[k];
                    int ny = j + dy[k];
                    if(nx >= 0 && nx < m && ny >= 0 && ny < n && grid[nx][ny] == '1')
                    {
                        int y = nx * n + ny;
                        unite.unite(x, y);
                    }
                }
            }
        }

        int ans = 0;
        for(int i = 0;i < m;i++)
        {
            for(int j = 0; j < n;j++)
            {
                if(grid[i][j] == '1')
                {
                    int x = i*n + j;
                    if(unite.find(x) == x)
                        ans++;
                }
            }
        }
        


    }
};

File: 练习/2025-12/岛屿数量.java
class Solution {
    public void dfs(char[][] grid, boolean[][] vis, int i, int j, int m, int n) {
        if (i < 0 || j < 0 || i >= m || j >= n || vis[i][j] || grid[i][j] == '0') {
            return;
        }
        vis[i][j] = true;
        dfs(grid, vis, i + 1, j, m, n);
        dfs(grid, vis, i - 1, j, m, n);
        dfs(grid, vis, i, j + 1, m, n);
        dfs(grid, vis, i, j - 1, m, n);
    }
    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        boolean[][] vis = new boolean[m][n];
        int ans = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (!vis[i][j] && grid[i][j] == '1') {
                    dfs(grid, vis, i, j, m, n);
                    ans++;
                }
            }
        }
        return ans;
    }
}